docs/building-deployer-envsubst.md

```diff
@@ -1,3 +1,226 @@
+`service` and `deployment` definition for nginx.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: my-nginx
+spec:
+  selector:
+    matchLabels:
+      run: my-nginx
+  replicas: $replicas
+  template:
+    metadata:
+      labels:
+        run: my-nginx
+    spec:
+      containers:
+      - name: my-nginx
+        image: nginx
+        ports:
+        - containerPort: 80
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-nginx
+  labels:
+    run: my-nginx
+spec:
+  ports:
+  - port: 80
+    protocol: TCP
+  selector:
+    run: my-nginx
+```
+
+### Build your deployer container
+
+Create a `Dockerfile` with the following content:
+
+```Dockerfile
+FROM gcr.io/cloud-marketplace-tools/k8s/deployer_envsubst/onbuild
+```
+
+Then you can build your container as follows:
+
+```bash
+# Set the registry to your project GCR repo.
+export REGISTRY=gcr.io/$(gcloud config get-value project | tr ':' '/')
+export APP_NAME=nginx
+
+docker build --tag $REGISTRY/$APP_NAME/deployer .
+```
+
+Push your built container to the remote GCR so that your app running in your GKE
+cluster can access the image:
+
+```dockerfile
+docker push $REGISTRY/$APP_NAME/deployer
+```
+
+### First deployment
+
+Create a new namespace to cleanly deploy your app:
+
+```shell
+kubectl create namespace test-ns
+
+mpdev /scripts/install \
+  --deployer=$REGISTRY/$APP_NAME/deployer \
+  --parameters='{"name": "test-deployment", "namespace": "test-ns", "replicas": 3 }'
+```
+
+See [mpdev reference](mpdev-references.md), for installing the mpdev tool.
+
+The `install` script simulates what the UI would do deploying your application.
+The app parameters are specified in a JSON string and `envsubst` is used to 
+substitute the parameters to the `*.template` files under the `manifest` 
+directory. In the UI, the user would have configured these parameters in a form.
+
+You can see your application in GKE UI by following this link:
+
+```text
+https://console.cloud.google.com/kubernetes/application?project=YOUR_PROJECT
+```
+
+## Part 2: Crafting the schema
+
+`schema.yaml` is a basic JSON schema with Marketplace extensions.
+See this [document](schema.md) for more references.
+
+### Parameterize the images
+
+Marketplace solutions must and can only use images that live on the official
+`marketplace.gcr.io`. Thus **all** images used in the manifests must be
+parameterized.
+
+Each image must have a corresponding property. At deployment time, the official
+images will be supplied via these properties.
+
+In the `schema.yaml` file, the images should look like this:
+```yaml
+# Under the x-google-marketplace parent attribute  
+  images:
+    nginx:
+      properties:
+        nginxImage:
+          type: FULL
+```
+
+The `manifests.yaml.template` file must be modified to reference `nginxImage`
+
+```yaml
+# Under spec.template
+    spec:
+      containers:
+      - name: my-nginx
+        image: $nginxImage # Changed from "nginx"
+```
+
```

Comment:
@huyhg I omitted the following lines from the helm deployer guide. Not sure if this still applies. If it does where are "default values" specified for images using `envsubst`?

https://github.com/GoogleCloudPlatform/marketplace-k8s-app-tools/blame/master/docs/building-deployer-helm.md#L406-L414